tfhe/src/shortint/server_key/tests/noise_level.rs

```diff
@@ -0,0 +1,288 @@
+use crate::shortint::ciphertext::NoiseLevel;
+use crate::shortint::keycache::KEY_CACHE;
+use crate::shortint::parameters::PARAM_MESSAGE_2_CARRY_2_KS_PBS;
+use crate::shortint::{Ciphertext, ServerKey};
+
+fn test_ct_unary_op_noise_level_propagation(sk: &ServerKey, ct: &Ciphertext) {
+    let test_fn = |op: &dyn Fn(&ServerKey, &Ciphertext) -> Ciphertext,
+                   predicate: &dyn Fn(NoiseLevel) -> NoiseLevel| {
+        assert!(op(sk, ct).noise_level() == predicate(ct.noise_level()));
+    };
+
+    test_fn(&ServerKey::unchecked_neg, &|ct_noise| ct_noise);
+    test_fn(
+        &|sk, ct| ServerKey::unchecked_neg_with_correcting_term(sk, ct).0,
+        &|ct_noise| ct_noise,
+    );
+
+    let acc = sk.generate_lookup_table(|_| 0);
+
+    test_fn(
+        &|sk, ct| ServerKey::apply_lookup_table(sk, ct, &acc),
+        &|_| NoiseLevel::NOMINAL,
+    );
+}
+
+fn test_ct_unary_op_assign_noise_level_propagation(sk: &ServerKey, ct: &Ciphertext) {
+    let test_fn = |op: &dyn Fn(&ServerKey, &mut Ciphertext),
+                   predicate: &dyn Fn(NoiseLevel) -> NoiseLevel| {
+        let mut clone = ct.clone();
+        op(sk, &mut clone);
+        assert!(clone.noise_level() == predicate(ct.noise_level()));
+    };
+
+    test_fn(&ServerKey::unchecked_neg_assign, &|ct_noise| ct_noise);
+    test_fn(
+        &|sk, ct| {
+            ServerKey::unchecked_neg_assign_with_correcting_term(sk, ct);
+        },
+        &|ct_noise| ct_noise,
+    );
+
+    let acc = sk.generate_lookup_table(|_| 0);
+
+    test_fn(
+        &|sk, ct| ServerKey::apply_lookup_table_assign(sk, ct, &acc),
+        &|_| NoiseLevel::NOMINAL,
+    );
+}
+
+fn test_ct_binary_op_noise_level_propagation(sk: &ServerKey, ct1: &Ciphertext, ct2: &Ciphertext) {
+    let test_fn = |op: &dyn Fn(&ServerKey, &Ciphertext, &Ciphertext) -> Ciphertext,
+                   predicate: &dyn Fn(NoiseLevel, NoiseLevel) -> NoiseLevel| {
+        assert!(op(sk, ct1, ct2).noise_level() == predicate(ct1.noise_level(), ct2.noise_level()));
+    };
+
+    test_fn(&ServerKey::unchecked_add, &|ct1_noise, ct2_noise| {
+        ct1_noise + ct2_noise
+    });
```

Comment:
I think you can easily add the default ops ?

Reply:
I plan to handle the noise level for operations (other than unchecked) and associated tests in subsequent PRs
In this PR, the noise level is updated but never used